app/models/bot/tagger.rb

```diff
@@ -0,0 +1,82 @@
+require 'json'
+
+class Bot::Tagger < BotUser
+  check_settings
+  class Error < ::StandardError
+  end
+
+  def self.get_tag_text(tag_id, auto_tag_prefix, ignore_autotags)
+    tag=TagText.find_by_id(tag_id)&.text
+    if tag.nil? || (ignore_autotags && tag[0]==auto_tag_prefix)
+      return nil
+    else
+      tag[0]==auto_tag_prefix ? tag[1..] : tag
+    end
+  end
+
+  def self.log(message, pm_id = nil, level = Logger::INFO)
+    prefix = "[AutoTagger Bot] "
+    prefix += "[ProjectMedia ##{pm_id}] " if pm_id
+    Rails.logger.log(level, "#{prefix} #{message}")
+  end
+
+  def self.run(body)
+    self.log("Received event with body of #{body}", nil, Logger::INFO)
+    if CheckConfig.get('alegre_host').blank?
+      self.log("Skipping events because `alegre_host` config is blank", nil, Logger::DEBUG)
+      return false
+    end
+
+    handled = false
+    pm = nil
+    begin
+      settings=JSON.parse(body[:settings])
+      auto_tag_prefix=settings["auto_tag_prefix"]
+      threshold=settings["threshold"]/100.0
+      ignore_autotags=settings["ignore_autotags"]
+      pm = ProjectMedia.where(id: body.dig(:data, :dbid)).last
+      if body.dig(:event) == 'create_project_media' && !pm.nil?
+        self.log("This item was just created, processing...", pm.id, Logger::INFO)
+        # Search all text fields for all items in the workspace using only the cofigured vector model
+        # The ProjectMedia's title is the query
+        # Do not use Elasticsearch. The threshold to use comes from the Tagger bot settings.
+        # Method signature: get_items_with_similar_text(pm, fields, threshold, query_text, models, team_ids = [pm&.team_id])
+        results=Bot::Alegre.get_items_with_similar_text(pm, Bot::Alegre::ALL_TEXT_SIMILARITY_FIELDS,
+          [{ value: threshold }], pm.title, [Bot::Alegre.matching_model_to_use(pm.team_id)].flatten.reject{|m| m==Bot::Alegre::ELASTICSEARCH_MODEL})
+        self.log("#{results.length} nearest neighbors #{results.keys()}", pm.id, Logger::INFO)
+        self.log("Results: #{results}", pm.id, Logger::INFO)
+
+        # For each nearest neighbor, get the tags.
+        tag_counts=results.map{|nn_pm,_| ProjectMedia.find(nn_pm).get_annotations('tag')}.flatten
+        # Transform from tag objects to strings and counts
+        tag_counts=tag_counts.map{|t| self.get_tag_text(t[:data][:tag],auto_tag_prefix,ignore_autotags)}.group_by(&:itself).transform_values(&:count)
+        # Reject any nil tags
+        tag_counts=tag_counts.reject{|k,_v|k==nil}.sort_by{|_k,v| v}
```

Comment:
**note:** here's a fun ruby shortcut - instead of`.reject{|k,_v|k==nil}` you can do `.compact`

Reply:
Nice. I'll check that out. Thank you.

Reply:
Looks like I'd have to put `compact` before the `group_by` as it doesn't remove key-value pairs
```
irb(main):015:0> fruits=["apple","apple","banana",nil,"raspberry","raspberry"]
=> ["apple", "apple", "banana", nil, "raspberry", "raspberry"]
irb(main):016:0> counts=fruits.group_by(&:itself).transform_values(&:count)
=> {"apple"=>2, "banana"=>1, nil=>1, "raspberry"=>2}
irb(main):017:0> counts.reject{|k,_v|k==nil}
=> {"apple"=>2, "banana"=>1, "raspberry"=>2}  #AS CURRENTLY IMPLEMENTED
irb(main):018:0> counts.compact
=> {"apple"=>2, "banana"=>1, nil=>1, "raspberry"=>2} # DOESN'T WORK
irb(main):019:0> fruits.compact.group_by(&:itself).transform_values(&:count)
=> {"apple"=>2, "banana"=>1, "raspberry"=>2} #WORKS
```